Add --force to delete to allow users to force delete whatever can be deleted

nikhiljindal opened the pull request:
Fixes https://github.com/GoogleCloudPlatform/k8s-multicluster-ingress/issues/164

Adding a `--force` flag to kubemci delete to continue deletion even in case of errors.

Will add an e2e test as part of remove-clusters test (delete the mci after it has been removed from all clusters) 

cc @G-Harmon

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/googlecloudplatform/k8s-multicluster-ingress/168)
<!-- Reviewable:end -->

G-Harmon commented:
mostly looks good. just a few minor comments.

---

Reviewed 3 of 3 files at r1.
Review status: all files reviewed at latest revision, all discussions resolved.

---

*[app/kubemci/cmd/delete.go, line 130 at r1](https://beta.reviewable.io/reviews/googlecloudplatform/k8s-multicluster-ingress/168#-L97omfR-zKgXKdxxyDh:-L97omfR-zKgXKdxxyDi:by7hbet) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/431169693ca1aab796809a6a180ecb0275e886ba/app/kubemci/cmd/delete.go#L130)):*
> ```Go
> 			return ingErr
> 		}
> 		fmt.Printf("%s. Continuing with force delete", ingErr)
> ```

Is it really a good idea to continue if UnmarshallAndApplyDefault fails? It just checks the namespaces and the annotations.

---

*[app/kubemci/cmd/delete.go, line 144 at r1](https://beta.reviewable.io/reviews/googlecloudplatform/k8s-multicluster-ingress/168#-L97p38p1Z9FFAEGzQO1:-L97p38p1Z9FFAEGzQO2:bdn51uv) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/431169693ca1aab796809a6a180ecb0275e886ba/app/kubemci/cmd/delete.go#L144)):*
> ```Go
> 			return clientsErr
> 		}
> 		fmt.Printf("%s. Continuing with force delete", clientsErr)
> ```

I think you need a newline at the end of Printf()s

---

*[app/kubemci/pkg/gcp/loadbalancer/loadbalancersyncer.go, line 231 at r1](https://beta.reviewable.io/reviews/googlecloudplatform/k8s-multicluster-ingress/168#-L97pLT6AUg2eOIrU0ml:-L97pLT6AUg2eOIrU0mm:b-avl7mv) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/431169693ca1aab796809a6a180ecb0275e886ba/app/kubemci/pkg/gcp/loadbalancer/loadbalancersyncer.go#L231)):*
> ```Go
> // DeleteLoadBalancer deletes the GCP resources associated with the L7 GCP load balancer for the given ingress.
> // TODO(nikhiljindal): Do not require the ingress yaml from users. Just the name should be enough. We can fetch ingress YAML from one of the clusters.
> func (l *LoadBalancerSyncer) DeleteLoadBalancer(ing *v1beta1.Ingress, forceDelete bool) error {
> ```

nit: document new parameters.

---

*[app/kubemci/pkg/gcp/loadbalancer/loadbalancersyncer_test.go, line 405 at r1](https://beta.reviewable.io/reviews/googlecloudplatform/k8s-multicluster-ingress/168#-L97pd7CDHh7ehRzevS0:-L97pd7DEt4N3J6cVubl:bcgdgxc) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/431169693ca1aab796809a6a180ecb0275e886ba/app/kubemci/pkg/gcp/loadbalancer/loadbalancersyncer_test.go#L405)):*
> ```Go
> 		t.Fatalf("unexpected error in test setup: %s", err)
> 	}
> 	if err := lbc.CreateLoadBalancer(ing, true /*forceUpdate*/, true /*validate*/, clusters); err != nil {
> ```

nit: set force=false while creating?

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/googlecloudplatform/k8s-multicluster-ingress/168#-:-L97qaBAAJoqM89TN2n-:b-taxokm)*
<!-- Sent from Reviewable.io -->

nikhiljindal commented:
Review status: 1 of 3 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[app/kubemci/cmd/delete.go, line 130 at r1](https://beta.reviewable.io/reviews/googlecloudplatform/k8s-multicluster-ingress/168#-L97omfR-zKgXKdxxyDh:-L9CdaaD07YIOnBdyR7o:b-tnrt3n) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/431169693ca1aab796809a6a180ecb0275e886ba/app/kubemci/cmd/delete.go#L130)):*
<details><summary><i>Previously, G-Harmon wrote…</i></summary><blockquote>

Is it really a good idea to continue if UnmarshallAndApplyDefault fails? It just checks the namespaces and the annotations.
</blockquote></details>

Right, removed. Thanks for pointing that out.

On second thought have also removed continuing from after GetClients. That also just instantiates clients based on kubeconfig.

---

*[app/kubemci/cmd/delete.go, line 144 at r1](https://beta.reviewable.io/reviews/googlecloudplatform/k8s-multicluster-ingress/168#-L97p38p1Z9FFAEGzQO1:-L9CdgNm85gOEGnAMole:bb6un1j) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/431169693ca1aab796809a6a180ecb0275e886ba/app/kubemci/cmd/delete.go#L144)):*
<details><summary><i>Previously, G-Harmon wrote…</i></summary><blockquote>

I think you need a newline at the end of Printf()s
</blockquote></details>

Done

---

*[app/kubemci/pkg/gcp/loadbalancer/loadbalancersyncer.go, line 231 at r1](https://beta.reviewable.io/reviews/googlecloudplatform/k8s-multicluster-ingress/168#-L97pLT6AUg2eOIrU0ml:-L9CeKUdBOcK_ajyQgS7:bb6un1j) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/431169693ca1aab796809a6a180ecb0275e886ba/app/kubemci/pkg/gcp/loadbalancer/loadbalancersyncer.go#L231)):*
<details><summary><i>Previously, G-Harmon wrote…</i></summary><blockquote>

nit: document new parameters.
</blockquote></details>

Done

---

*[app/kubemci/pkg/gcp/loadbalancer/loadbalancersyncer_test.go, line 405 at r1](https://beta.reviewable.io/reviews/googlecloudplatform/k8s-multicluster-ingress/168#-L97pd7CDHh7ehRzevS0:-L9Ce_Ia7etgLj-UOPFj:b8t23cx) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/431169693ca1aab796809a6a180ecb0275e886ba/app/kubemci/pkg/gcp/loadbalancer/loadbalancersyncer_test.go#L405)):*
<details><summary><i>Previously, G-Harmon wrote…</i></summary><blockquote>

nit: set force=false while creating?
</blockquote></details>

Right. Good catch

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/googlecloudplatform/k8s-multicluster-ingress/168)*
<!-- Sent from Reviewable.io -->

nikhiljindal commented:
Thanks for the review @G-Harmon 
Updated code as per comments

G-Harmon commented:
Reviewed 1 of 2 files at r2, 2 of 3 files at r3.
Review status: 2 of 3 files reviewed at latest revision, 1 unresolved discussion.

---

*[app/kubemci/cmd/delete.go, line 130 at r1](https://beta.reviewable.io/reviews/googlecloudplatform/k8s-multicluster-ingress/168#-L97omfR-zKgXKdxxyDh:-L9D4Wb9Au-bjvUSLn-8:b-z04cta) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/431169693ca1aab796809a6a180ecb0275e886ba/app/kubemci/cmd/delete.go#L130)):*
<details><summary><i>Previously, nikhiljindal (Nikhil Jindal) wrote…</i></summary><blockquote>

Right, removed. Thanks for pointing that out.

On second thought have also removed continuing from after GetClients. That also just instantiates clients based on kubeconfig.
</blockquote></details>

Did you actually change and push this? it looks the same...

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/googlecloudplatform/k8s-multicluster-ingress/168)*
<!-- Sent from Reviewable.io -->

nikhiljindal commented:
Review status: 2 of 3 files reviewed at latest revision, 1 unresolved discussion.

---

*[app/kubemci/cmd/delete.go, line 130 at r1](https://beta.reviewable.io/reviews/googlecloudplatform/k8s-multicluster-ingress/168#-L97omfR-zKgXKdxxyDh:-L9DAVupDKwU7rf02823:b-ku7ezr) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/431169693ca1aab796809a6a180ecb0275e886ba/app/kubemci/cmd/delete.go#L130)):*
<details><summary><i>Previously, G-Harmon wrote…</i></summary><blockquote>

Did you actually change and push this? it looks the same...
</blockquote></details>

Yes I did push the change.  If you go directly to https://beta.reviewable.io/reviews/googlecloudplatform/k8s-multicluster-ingress/168 then you will see the change. If you click on the link for this comment, it takes you to https://beta.reviewable.io/reviews/googlecloudplatform/k8s-multicluster-ingress/168#-L97omfR-zKgXKdxxyDh, which shows an older revision (the one with the comment).

Maybe you clicked on the link and hence are viewing an older revision?

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/googlecloudplatform/k8s-multicluster-ingress/168)*
<!-- Sent from Reviewable.io -->

G-Harmon commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 3 files at r3.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[app/kubemci/cmd/delete.go, line 130 at r1](https://beta.reviewable.io/reviews/googlecloudplatform/k8s-multicluster-ingress/168#-L97omfR-zKgXKdxxyDh:-L9DzlpP7ymY9kWTGDEu:b-yfd5i0) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/431169693ca1aab796809a6a180ecb0275e886ba/app/kubemci/cmd/delete.go#L130)):*
<details><summary><i>Previously, nikhiljindal (Nikhil Jindal) wrote…</i></summary><blockquote>

Yes I did push the change.  If you go directly to https://beta.reviewable.io/reviews/googlecloudplatform/k8s-multicluster-ingress/168 then you will see the change. If you click on the link for this comment, it takes you to https://beta.reviewable.io/reviews/googlecloudplatform/k8s-multicluster-ingress/168#-L97omfR-zKgXKdxxyDh, which shows an older revision (the one with the comment).

Maybe you clicked on the link and hence are viewing an older revision?
</blockquote></details>

I thought I had changed it to look at the latest rev, but I think I was actually comparing r1 to r2. oops.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/googlecloudplatform/k8s-multicluster-ingress/168#-:-L9E-W4dDF25vQE2LLOP:bnfp4nl)*
<!-- Sent from Reviewable.io -->
